backend/cfg/eliminate_dead_blocks.ml

```diff
@@ -64,7 +64,13 @@ let rec eliminate_dead_blocks cfg_with_layout =
       Printf.printf "\n" );
     (* Termination: the number of remaining blocks is strictly smaller in
        each recursive call. *)
-    eliminate_dead_blocks cfg_with_layout )
+    eliminate_dead_blocks cfg_with_layout;
+    begin match cfg.fun_tailrec_entry_point_label with
+    | None -> ()
+    | Some label ->
+      if List.exists (Label.equal label) found_dead then
+        Cfg.set_fun_tailrec_entry_point_label cfg None
+    end)
```

Comment:
I think this check is not correct here: if there is another `Tailcall Self` that is not dead, we cannot set `fun_tailrec_entry_point_label` to `None`. Instead, we should check at the end of CFG construction and transformations whether any block has Tailcall Self and if not, then we can set `fun_tailrec_entry_point_label` to None.

Reply:
sorry, this is actually fine and I was confused about who is dead! I still think it's better to do this check at the end.

Reply:
> I still think it's better to do this check at the end.

I am slightly confused: it looks like the operation is the very
last of the function, and hence the whole transformation.

Reply:
This check will happen at the end of each recursive call.
What I meant is to have a separate pass that checks if the input Cfg has any Tailcall Self, and if it does not, sets tailrec label to None. This pass can be called at the end of Linear_to_cfg.run, and at the end of Eliminate_dead_blocks.run and at the end of Eliminate_fallthrough_blocks.run.
What you have here is more efficient, because it fixes up tailrec label only as needed, but I think we have this problem in  the first place because the representation of tailrec label is problematic. 

As suggested offline, this problem would completely go away if we change the representation of Tailcall Self to include the label, and also remove fun_tailrec_entry_point_label in Cfg. It is still worth changing Linear.fundecl.fun_tailrec_entry_point_label to an option. The pass Cfg_to_linear can compute the field and check its consistency (i.e., all Tailrec Self have the same destination label).

Reply:
I understand your point about the recursive calls.
However, it sounds simpler to have these calls accumulate
the set of dead blocks rather than have a whole new phase/
traversal at the end of `run`. `run` would get the set of
eliminated blocks from `eliminate_dead_blocks` and the
code currently at lines 68-73 would be at the end of `run`.